code/iaas/allocator/src/main/java/io/cattle/platform/allocator/service/AbstractAllocator.java

```diff
@@ -228,13 +239,58 @@ protected boolean doAllocate(final AllocationRequest request, final AllocationAt
 
     protected boolean acquireLockAndAllocate(final AllocationRequest request, final AllocationAttempt attempt, Object deallocate) {
         final List<Constraint> finalFailedConstraints = new ArrayList<>();
+        final List<AllocationCandidate> candidatesForCPUMemoryIops = new ArrayList<AllocationCandidate>();
+        final List<AllocationCandidate> candidates = new ArrayList<AllocationCandidate>();
         lockManager.lock(getAllocationLock(request, attempt), new LockCallbackNoReturn() {
             @Override
             public void doWithLockNoResult() {
                 Context c = allocateTimer.time();
                 try {
+                    Iterator<AllocationCandidate> iter = getCandidates(attempt);
+                    while (iter.hasNext()) {
+                        candidatesForCPUMemoryIops.add(iter.next());
+                    }
+                    if (iter != null) {
+                        close(iter);
+                    }
+
+                    Set<Long> alreadyScheduledHostIds = new HashSet<Long>();
+
+                    // scheduler the list and return a list that is able to schedule
+                    for (AllocationCandidate candidate : candidatesForCPUMemoryIops) {
+                        boolean good = true;
+                        
+                        // schedule cpu, memory and iops first
+                        try {
+                            for( Long hostId : candidate.getHosts() ){
+
+                                // in case we already tried to schedule instance for this hostId, then skip it
+                                if (alreadyScheduledHostIds.contains(hostId)) {
+                                    continue;
```

Comment:
I think this might need to change this logic.

I hit a bug where two candidates containing the same hostId were passed in. They should not have been allowed to schedule to that I was assigning them to because the host didnt have fio data and the iops labels were set.

For the first host, the scheduler reported no and we hit the continue on line 286. Then the next candidate went. good was set back to true. then we hit the alreadyScheduledHostIds.contains(hostId) logic and it broke out of the loop without ever setting good to false. So, the host was used